### PR TITLE
Convert utils.js -> utils.ts

### PR DESCRIPTION
**Pull request by foldleft**
Note: if you set `noImplicitAny` to `true` this still compiles :)

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -20,15 +20,15 @@ limitations under the License.
...
+import * as unhomoglyph from 'unhomoglyph';
 
 /**
  * Encode a dictionary of query parameters.
  * @param {Object} params A dict of key/values to encode e.g.
  * {"foo": "bar", "baz": "taz"}
  * @return {string} The encoded string e.g. foo=bar&baz=taz
  */
-export function encodeParams(params) {
+export function encodeParams(params: {[key: string]: string}): string {
```
Comment (t3chguy) [inline src/utils.ts]: could we use the more readable utility type here? `Record<string, string>`

File: src/utils.ts [hunk 2]
```diff
@@ -48,7 +48,8 @@ export function encodeParams(params) {
  * variables with. E.g. { "$bar": "baz" }.
  * @return {string} The result of replacing all template variables e.g. '/foo/baz'.
  */
-export function encodeUri(pathTemplate, variables) {
+export function encodeUri(pathTemplate: string,
+                          variables: {[key: string]: string}): string {
```
Comment (t3chguy) [inline src/utils.ts]: same as above

File: src/utils.ts [hunk 3]
```diff
@@ -114,7 +116,7 @@ export function keys(obj) {
  * @param {Object} obj The object to get the values for.
  * @return {Array<*>} The values of the object.
  */
-export function values(obj) {
+export function values<T>(obj: {[keys: string]: T}): T[] {
```
Comment (t3chguy) [inline src/utils.ts]: same as above

File: src/utils.ts (hunk 1 above)
Comment (t3chguy) [reply src/utils.ts]: https://www.typescriptlang.org/docs/handbook/utility-types.html

**Comment by foldleft** [comment]
ooh, I did not know about those! Thanks :)

**Comment by turt2live** [comment]
(you also have build failures to look at)

File: src/@types/unhomoglyph/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+declare function unhomoglyph (s: string): string;
+
+declare module 'unhomoglyph' {
```
Comment (turt2live) [inline src/@types/unhomoglyph/index.d.ts]: We shouldn't bother with third party declarations in our projects.

File: src/utils.ts [hunk 5]
```diff
@@ -617,7 +628,7 @@ export function inherits(ctor, superCtor) {
  * @param {any} SuperType The type to act as a super instance
  * @param {any} params Arguments to supply to the super type's constructor
  */
-export function polyfillSuper(thisArg, SuperType, ...params) {
+export function polyfillSuper(thisArg: any, SuperType: any, ...params: any) {
```
Comment (turt2live) [inline src/utils.ts]:
```suggestion
export function polyfillSuper(thisArg: any, SuperType: any, ...params: any[]) {
```

File: tsconfig.json [hunk 6]
```diff
@@ -5,6 +5,7 @@
     "module": "commonjs",
     "moduleResolution": "node",
     "target": "es2016",
+    "esModuleInterop": true,
```
Comment (turt2live) [inline tsconfig.json]: from the docs, we shouldn't need this?

File: src/utils.ts (hunk 5 above)
Comment (turt2live) [reply src/utils.ts]: this is because the spread operator is an array btw

**Pull request merged**